### PR TITLE
FIX: Allow default UI actions to work without setting up a project-wide actions file (ISXB-811).

### PR DESCRIPTION
**Pull request by graham-huws**
### Description

Regression from previous versions: Default UI interactions like clicking, dragging etc should work without users having to set up their own PWA file. [ISXB-811](https://jira.unity3d.com/browse/ISXB-811). Lengthy discussion in Slack [here](https://unity.slack.com/archives/C04RDRXRJ1L/p1711549077482879).

### Changes made

Load DefaultInputActions file if a project-wide action has not been set by the user. Enable the UI functions from it.
Also, fix newlines not functioning correctly in a uxml file - details [here](https://unity.slack.com/archives/C3414V4UV/p1673359236696379?thread_ts=1673347662.674689&cid=C3414V4UV).

### Checklist

Before review:

- [x] Changelog entry added.
    - Explains the change in `Changed`, `Fixed`, `Added` sections.
    - For API change contains an example snippet and/or migration example.
    - FogBugz ticket attached, example `([case %number%](https://issuetracker.unity3d.com/issues/...))`.
    - FogBugz is marked as "Resolved" with *next* release version correctly set.
- [ ] Tests added/changed, if applicable.
    - Functional tests `Area_CanDoX`, `Area_CanDoX_EvenIfYIsTheCase`, `Area_WhenIDoX_AndYHappens_ThisIsTheResult`.
    - Performance tests.
    - Integration tests.
- [ ] Docs for new/changed API's.
    - Xmldoc cross references are set correctly.
    - Added explanation how the API works.
    - Usage code examples added.
    - The manual is updated, if needed.

During merge:

- [x] Commit message for squash-merge is prefixed with one of the list:
    - `NEW: ___`.
    - `FIX: ___`.
    - `DOCS: ___`.
    - `CHANGE: ___`.
    - `RELEASE: 1.1.0-preview.3`.

### CODE REVIEW

**Review by jfreire-unity — CHANGES_REQUESTED** [review]
Code looks ok to me 👍
But I would like see at least a small test to make sure the problem is solved. Essentially, something that shows that these default actions are working if there is no project wide actions asset available.

**Comment by jfreire-unity** [comment]
As discussed with @graham-huws , I'll add this test while he focus on a template change fix.

**Comment by unity-cla-assistant** [comment]
[CLA assistant check](https://cla-assistant.cds.internal.unity3d.com/Unity-Technologies/InputSystem?pullRequest=1899) <br/>All committers have signed the CLA.

**Review by jfreire-unity — APPROVED** [review]

File: Assets/Tests/InputSystem/Plugins/InputForUITests.cs
Comment (ekcoh) [inline Assets/Tests/InputSystem/Plugins/InputForUITests.cs]: I would suggest to not have asserts in Setup and only have asserts to actually verify functional requirements as part of test methods. For what scenario would this fail were it wouldn't fail the functional tests anyway? Otherwise if it renders the test non-applicable use assumption failure or throw exception if programmer error?

**Comment by Pauliusd01** [comment]
I'm getting this error when ProjectWideActions asset is used but does not contain the UI map:

```
NullReferenceException: Object reference not set to an instance of an object
UnityEngine.InputSystem.Plugins.InputForUI.InputSystemProvider.ReadCurrentNavigationMoveVector () (at ./Packages/com.unity.inputsystem/InputSystem/Plugins/InputForUI/InputSystemProvider.cs:215)
UnityEngine.InputSystem.Plugins.InputForUI.InputSystemProvider.DirectionNavigation (Unity.IntegerTime.DiscreteTime currentTime) (at ./Packages/com.unity.inputsystem/InputSystem/Plugins/InputForUI/InputSystemProvider.cs:164)
UnityEngine.InputSystem.Plugins.InputForUI.InputSystemProvider.Update () (at ./Packages/com.unity.inputsystem/InputSystem/Plugins/InputForUI/InputSystemProvider.cs:127)
UnityEngine.InputForUI.EventProvider.NotifyUpdate () (at D:/Gitrepo/unity/Modules/InputForUI/Provider/EventProvider.cs:164)
```
Previously I would get a warning telling me that there will be issues unless UI map is restored. Is this expected?

**Comment by Pauliusd01** [comment]
I am also getting this error after resetting the ProjectWideActions asset from having no UI map to its defaults:

```
InvalidOperationException: Action 'UI/Point[/Mouse/position]' must be part of an InputActionAsset in order to be able to create an InputActionReference for it
UnityEngine.InputSystem.InputActionReference.Set (UnityEngine.InputSystem.InputAction action) (at ./Packages/com.unity.inputsystem/InputSystem/Actions/InputActionReference.cs:92)
UnityEngine.InputSystem.InputActionReference.Create (UnityEngine.InputSystem.InputAction action) (at ./Packages/com.unity.inputsystem/InputSystem/Actions/InputActionReference.cs:198)
UnityEngine.InputSystem.Plugins.InputForUI.InputSystemProvider.RegisterActions (UnityEngine.InputSystem.Plugins.InputForUI.InputSystemProvider+Configuration cfg) (at ./Packages/com.unity.inputsystem/InputSystem/Plugins/InputForUI/InputSystemProvider.cs:582)
UnityEngine.InputSystem.Plugins.InputForUI.InputSystemProvider.Initialize () (at ./Packages/com.unity.inputsystem/InputSystem/Plugins/InputForUI/InputSystemProvider.cs:84)
UnityEngine.InputForUI.EventProvider.Initialize () (at D:/Gitrepo/unity/Modules/InputForUI/Provider/EventProvider.cs:125)
UnityEngine.InputForUI.EventProvider.SetEnabled (System.Boolean enable) (at D:/Gitrepo/unity/Modules/InputForUI/Provider/EventProvider.cs:60)
UnityEngine.UIElements.DefaultEventSystem.InitInputProcessor () (at D:/Gitrepo/unity/Modules/UIElements/Core/DefaultEventSystem.cs:163)
UnityEngine.UIElements.DefaultEventSystem.set_isInputReady (System.Boolean value) (at D:/Gitrepo/unity/Modules/UIElements/Core/DefaultEventSystem.cs:94)
UnityEngine.UIElements.UIElementsRuntimeUtility.RegisterPlayerloopCallback () (at D:/Gitrepo/unity/Modules/UIElements/Core/UIElementsRuntimeUtility.cs:248)
UnityEngine.UIElements.UIElementsRuntimeUtility.RegisterCachedPanelInternal (System.Int32 instanceID, UnityEngine.UIElements.IPanel panel) (at D:/Gitrepo/unity/Modules/UIElements/Core/UIElementsRuntimeUtility.cs:61)
UnityEngine.UIElements.UIElementsRuntimeUtility.FindOrCreateRuntimePanel (UnityEngine.ScriptableObject ownerObject, UnityEngine.UIElements.UIElementsRuntimeUtility+CreateRuntimePanelDelegate createDelegate) (at D:/Gitrepo/unity/Modules/UIElements/Core/UIElementsRuntimeUtility.cs:37)
UnityEngine.UIElements.PanelSettings+RuntimePanelAccess.CreateRelatedRuntimePanel () (at D:/Gitrepo/unity/Modules/UIElements/Core/GameObjects/PanelSettings.cs:160)
UnityEngine.UIElements.PanelSettings+RuntimePanelAccess.get_panel () (at D:/Gitrepo/unity/Modules/UIElements/Core/GameObjects/PanelSettings.cs:91)
UnityEngine.UIElements.PanelSettings.get_panel () (at D:/Gitrepo/unity/Modules/UIElements/Core/GameObjects/PanelSettings.cs:544)
UnityEngine.UIElements.UIDocument.SetupVisualTreeAssetTracker () (at D:/Gitrepo/unity/Modules/UIElements/Core/GameObjects/UIDocument.cs:823)
UnityEngine.UIElements.UIDocument.RecreateUI () (at D:/Gitrepo/unity/Modules/UIElements/Core/GameObjects/UIDocument.cs:619)
UnityEngine.UIElements.UIDocument.set_visualTreeAsset (UnityEngine.UIElements.VisualTreeAsset value) (at D:/Gitrepo/unity/Modules/UIElements/Core/GameObjects/UIDocument.cs:227)
UnityEngine.UIElements.UIDocument.OnValidate () (at D:/Gitrepo/unity/Modules/UIElements/Core/GameObjects/UIDocument.cs:912)
```

File: Assets/Tests/InputSystem/Plugins/InputForUITests.cs
Comment (ekcoh) [inline Assets/Tests/InputSystem/Plugins/InputForUITests.cs:169]: I see this bypasses the play mode check by accessing internals directly, seems like a pragmatic solution that eliminates the need to mock via IInputRuntime interface discussed previously. Just maybe note that this is not representative of what happens in reality since this can only can happen in edit mode.

I assume the intention here is to test same behavior with and without PWA. However, this should be applicable for all tests in this class, and hence I would recommend either parameterising all test cases in this class or use two separate test classes that inherits the test methods and use different fixed setups instead?

**Comment by Pauliusd01** [comment]
> I'm getting this error when ProjectWideActions asset is used but does not contain the UI map:
> 
> ```
> NullReferenceException: Object reference not set to an instance of an object
> UnityEngine.InputSystem.Plugins.InputForUI.InputSystemProvider.ReadCurrentNavigationMoveVector () (at ./Packages/com.unity.inputsystem/InputSystem/Plugins/InputForUI/InputSystemProvider.cs:215)
> UnityEngine.InputSystem.Plugins.InputForUI.InputSystemProvider.DirectionNavigation (Unity.IntegerTime.DiscreteTime currentTime) (at ./Packages/com.unity.inputsystem/InputSystem/Plugins/InputForUI/InputSystemProvider.cs:164)
> UnityEngine.InputSystem.Plugins.InputForUI.InputSystemProvider.Update () (at ./Packages/com.unity.inputsystem/InputSystem/Plugins/InputForUI/InputSystemProvider.cs:127)
> UnityEngine.InputForUI.EventProvider.NotifyUpdate () (at D:/Gitrepo/unity/Modules/InputForUI/Provider/EventProvider.cs:164)
> ```
> 
> Previously I would get a warning telling me that there will be issues unless UI map is restored. Is this expected?



> I am also getting this error after resetting the ProjectWideActions asset from having no UI map to its defaults:
> 
> ```
> InvalidOperationException: Action 'UI/Point[/Mouse/position]' must be part of an InputActionAsset in order to be able to create an InputActionReference for it
> UnityEngine.InputSystem.InputActionReference.Set (UnityEngine.InputSystem.InputAction action) (at ./Packages/com.unity.inputsystem/InputSystem/Actions/InputActionReference.cs:92)
> UnityEngine.InputSystem.InputActionReference.Create (UnityEngine.InputSystem.InputAction action) (at ./Packages/com.unity.inputsystem/InputSystem/Actions/InputActionReference.cs:198)
> UnityEngine.InputSystem.Plugins.InputForUI.InputSystemProvider.RegisterActions (UnityEngine.InputSystem.Plugins.InputForUI.InputSystemProvider+Configuration cfg) (at ./Packages/com.unity.inputsystem/InputSystem/Plugins/InputForUI/InputSystemProvider.cs:582)
> UnityEngine.InputSystem.Plugins.InputForUI.InputSystemProvider.Initialize () (at ./Packages/com.unity.inputsystem/InputSystem/Plugins/InputForUI/InputSystemProvider.cs:84)
> UnityEngine.InputForUI.EventProvider.Initialize () (at D:/Gitrepo/unity/Modules/InputForUI/Provider/EventProvider.cs:125)
> UnityEngine.InputForUI.EventProvider.SetEnabled (System.Boolean enable) (at D:/Gitrepo/unity/Modules/InputForUI/Provider/EventProvider.cs:60)
> UnityEngine.UIElements.DefaultEventSystem.InitInputProcessor () (at D:/Gitrepo/unity/Modules/UIElements/Core/DefaultEventSystem.cs:163)
> UnityEngine.UIElements.DefaultEventSystem.set_isInputReady (System.Boolean value) (at D:/Gitrepo/unity/Modules/UIElements/Core/DefaultEventSystem.cs:94)
> UnityEngine.UIElements.UIElementsRuntimeUtility.RegisterPlayerloopCallback () (at D:/Gitrepo/unity/Modules/UIElements/Core/UIElementsRuntimeUtility.cs:248)
> UnityEngine.UIElements.UIElementsRuntimeUtility.RegisterCachedPanelInternal (System.Int32 instanceID, UnityEngine.UIElements.IPanel panel) (at D:/Gitrepo/unity/Modules/UIElements/Core/UIElementsRuntimeUtility.cs:61)
> UnityEngine.UIElements.UIElementsRuntimeUtility.FindOrCreateRuntimePanel (UnityEngine.ScriptableObject ownerObject, UnityEngine.UIElements.UIElementsRuntimeUtility+CreateRuntimePanelDelegate createDelegate) (at D:/Gitrepo/unity/Modules/UIElements/Core/UIElementsRuntimeUtility.cs:37)
> UnityEngine.UIElements.PanelSettings+RuntimePanelAccess.CreateRelatedRuntimePanel () (at D:/Gitrepo/unity/Modules/UIElements/Core/GameObjects/PanelSettings.cs:160)
> UnityEngine.UIElements.PanelSettings+RuntimePanelAccess.get_panel () (at D:/Gitrepo/unity/Modules/UIElements/Core/GameObjects/PanelSettings.cs:91)
> UnityEngine.UIElements.PanelSettings.get_panel () (at D:/Gitrepo/unity/Modules/UIElements/Core/GameObjects/PanelSettings.cs:544)
> UnityEngine.UIElements.UIDocument.SetupVisualTreeAssetTracker () (at D:/Gitrepo/unity/Modules/UIElements/Core/GameObjects/UIDocument.cs:823)
> UnityEngine.UIElements.UIDocument.RecreateUI () (at D:/Gitrepo/unity/Modules/UIElements/Core/GameObjects/UIDocument.cs:619)
> UnityEngine.UIElements.UIDocument.set_visualTreeAsset (UnityEngine.UIElements.VisualTreeAsset value) (at D:/Gitrepo/unity/Modules/UIElements/Core/GameObjects/UIDocument.cs:227)
> UnityEngine.UIElements.UIDocument.OnValidate () (at D:/Gitrepo/unity/Modules/UIElements/Core/GameObjects/UIDocument.cs:912)
> ```

Repro'd both on develop, disregard them. (They are filed separately: https://jira.unity3d.com/browse/ISX-1953 / https://jira.unity3d.com/browse/ISX-1954 )

File: Packages/com.unity.inputsystem/CHANGELOG.md
Comment (ekcoh) [inline Packages/com.unity.inputsystem/CHANGELOG.md]: I would recommend rewriting this in the format that explains the problem+symptom and new behavior. E.g. (not sure if below is 100% correct, but just to illustrate what I mean)

Fixes an issue where UGUI or UI Toolkit would not react to input unless a Project-wide Input Action Asset was configured for the project. If no Project-wide Input Action Asset has been configured for the project, default actions are now instead created on-the-fly to provide default bindings for UI actions.

File: Packages/com.unity.inputsystem/InputSystem/Editor/UITKAssetEditor/Resources/InputActionsProjectSettings.uxml
Comment (ekcoh) [inline Packages/com.unity.inputsystem/InputSystem/Editor/UITKAssetEditor/Resources/InputActionsProjectSettings.uxml:11]: Is this an additional fix not called out or a reformatting mistake?

File: Packages/com.unity.inputsystem/InputSystem/Plugins/InputForUI/InputSystemProvider.cs
Comment (ekcoh) [inline Packages/com.unity.inputsystem/InputSystem/Plugins/InputForUI/InputSystemProvider.cs]: Why did this change to public visibility?

**Review by ekcoh — COMMENTED** [review]
Thanks for submitting this PR. Fix looks good since it avoids regressive behavior even if it might imply differentiation between PWA and DefaultActions bindings. 

I still think this needs more test coverage. I would like to see that all UI actions have associated test cases (may happen outside this PR but would not suggest delaying). We may not necessarily cover all bindings (but it would be good for defaults) but if we do not cover all actions this may break without us noticing. Hence I would recommend adding test coverage for all actions.

In addition it would be good to add tests that covers:
- Behavior when assigned PWA asset is modified after being assigned to InputSystemProvider. 
- Behavior when assigned PWA asset is not containing required actions at all.
- Behavior when assigned PWA asset is partially containing required actions.
- Behavior when assigned PWA asset is containing required actions but with invalid action types.
- Behavior when assigned PWA asset is containing required actions but has no bindings on them.

I am not sure if this PR is intended to cover the task of improving the UI integration test coverage but if it is I would recommend covering tests above. If desired to handle outside this PR to land fix this is OK but then test coverage shouldn't be postponed. Feel free to reach out directly for further discussion.

**Review by ekcoh — CHANGES_REQUESTED** [review]
See previous comments on test coverage (seem to also have problems in the area found by Paulius). I am open to handle parts of the test coverage outside this PR if it reduces delay, but do not suggest postponing it. Also have some open questions.

File: Packages/com.unity.inputsystem/InputSystem/Editor/UITKAssetEditor/Resources/InputActionsProjectSettings.uxml
Comment (graham-huws-unity) [reply Packages/com.unity.inputsystem/InputSystem/Editor/UITKAssetEditor/Resources/InputActionsProjectSettings.uxml:11]: This is called out in "Changes made" in the PR description.

**Review by Pauliusd01 — APPROVED** [review]
LGTM. Checked UITK functionality with ProjectWideActions asset and without, in different backend modes (Old, New, Both) and with/without EventSystem object - works fine now.

I do think that https://jira.unity3d.com/browse/ISX-1953 / https://jira.unity3d.com/browse/ISX-1954 need to be fixed for it to be in a good state though but they are not caused by this PR.

File: Packages/com.unity.inputsystem/InputSystem/Plugins/InputForUI/InputSystemProvider.cs
Comment (benoitalain) [inline Packages/com.unity.inputsystem/InputSystem/Plugins/InputForUI/InputSystemProvider.cs:586]: We're passing `cfg` as an argument but we're using `m_Cfg`. Either don't pass `cfg` at all or use it instead of the global variable.

File: Packages/com.unity.inputsystem/InputSystem/Plugins/InputForUI/InputSystemProvider.cs
Comment (benoitalain) [inline Packages/com.unity.inputsystem/InputSystem/Plugins/InputForUI/InputSystemProvider.cs:621]: OK. Maybe we should only create the UI actions, instead of creating all and leaving the rest disabled? But I understand that this would make the code more error-prone and isn't necessary at all for the current fix.

**Review by benoitalain — APPROVED** [review]
I tested the fix and it works.

File: Assets/Tests/InputSystem/Plugins/InputForUITests.cs
Comment (benoitalain) [inline Assets/Tests/InputSystem/Plugins/InputForUITests.cs:155]: I have this one failing locally right now.

File: Assets/Tests/InputSystem/Plugins/InputForUITests.cs
Comment (benoitalain) [reply Assets/Tests/InputSystem/Plugins/InputForUITests.cs:169]: Also we need to not corrupt the state of `InputSystem.s_Manager.actions` after running this test. I would recommend either storing its value locally and setting it back in a `finally` block, or making the state be reset in `TearDown`.

Comment (ekcoh) [reply Assets/Tests/InputSystem/Plugins/InputForUITests.cs:169]: @benoitalain This was also my first thought, but I believe its not corrupted, its automatically reinitialised as part of InputTestFixture reinitialisation.

File: Packages/com.unity.inputsystem/InputSystem/Editor/UITKAssetEditor/Resources/InputActionsProjectSettings.uxml
Comment (ekcoh) [reply Packages/com.unity.inputsystem/InputSystem/Editor/UITKAssetEditor/Resources/InputActionsProjectSettings.uxml:11]: Sorry, guess I overlooked it

File: Assets/Tests/InputSystem/Plugins/InputForUITests.cs
Comment (graham-huws-unity) [reply Assets/Tests/InputSystem/Plugins/InputForUITests.cs:155]: Have been discussing with @Pauliusd01 and @ekcoh - There's an issue with the way input tests are setup which makes their behaviour inconsistent, which needs addressing as a separate issue. For example, running the whole of the "Tests.dll" playmode section makes this test pass, but running just the sub-group "InputForUITests" has this test - and on another run, the rest in that group - fail. I don't think it's worth holding this PR up, as all manual testing shows the change is working as intended.

File: Packages/com.unity.inputsystem/InputSystem/Plugins/InputForUI/InputSystemProvider.cs
Comment (graham-huws-unity) [reply Packages/com.unity.inputsystem/InputSystem/Plugins/InputForUI/InputSystemProvider.cs]: Updated in e99ed66a5b8bc1887f3527cf35d03b809e34d548 to be more restrictive.

File: Packages/com.unity.inputsystem/InputSystem/Plugins/InputForUI/InputSystemProvider.cs
Comment (graham-huws-unity) [reply Packages/com.unity.inputsystem/InputSystem/Plugins/InputForUI/InputSystemProvider.cs:621]: Yeah, that was my line of thinking - plus, this will need a fuller re-write in the future anyway to remove DefaultInputActions.

File: Packages/com.unity.inputsystem/InputSystem/Plugins/InputForUI/InputSystemProvider.cs
Comment (graham-huws-unity) [reply Packages/com.unity.inputsystem/InputSystem/Plugins/InputForUI/InputSystemProvider.cs:586]: Fix in 947912723ac498e31c36c4f0c63c04a5c1712b9e.

File: Packages/com.unity.inputsystem/CHANGELOG.md
Comment (graham-huws-unity) [reply Packages/com.unity.inputsystem/CHANGELOG.md]: Tweaked this in 982b8abdd43d4e7f05fa560d305f2f5f972c6fef.

File: Assets/Tests/InputSystem/Plugins/InputForUITests.cs
Comment (graham-huws-unity) [reply Assets/Tests/InputSystem/Plugins/InputForUITests.cs]: I've moved this check out to a separate test in 404f0a2b9aefb2d21e7e69038f0103a5f5167189.

**Review by ekcoh — APPROVED** [review]

**Pull request merged**
